warehouse/oidc/forms/gitlab.py

```diff
@@ -19,11 +19,117 @@
 from warehouse.oidc.forms._core import PendingPublisherMixin
 
 # https://docs.gitlab.com/ee/user/reserved_names.html#limitations-on-project-and-group-names
-_VALID_GITLAB_PROJECT = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9-_.]*$")
-_VALID_GITLAB_NAMESPACE = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9-_./]*$")
+_VALID_GITLAB_PROJECT = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9_.-]*[a-zA-Z0-9]$")
+_VALID_GITLAB_NAMESPACE = re.compile(r"^[a-zA-Z0-9][a-zA-Z0-9-_./+]*[a-zA-Z0-9]$")
 _VALID_GITLAB_ENVIRONMENT = re.compile(r"^[a-zA-Z0-9\-_/${} ]+$")
 
+_GITLAB_SPECIAL_CHARACTERS = [".", "-", "_"]
+
+_GITLAB_RESERVED_PROJECT_NAMES = {
+    "-",
+    "badges",
+    "blame",
+    "blob",
+    "builds",
+    "commits",
+    "create",
+    "create_dir",
+    "edit",
+    "environments/folders",
+    "files",
+    "find_file",
+    "gitlab-lfs/objects",
+    "info/lfs/objects",
+    "new",
+    "preview",
+    "raw",
+    "refs",
+    "tree",
+    "update",
+    "wikis",
+}
+
+_GITLAB_RESERVED_GROUP_NAMES = {
+  "-",
+  ".well-known",
+  "404.html",
+  "422.html",
+  "500.html",
+  "502.html",
+  "503.html",
+  "admin",
+  "api",
+  "apple-touch-icon.png",
+  "assets",
+  "dashboard",
+  "deploy.html",
+  "explore",
+  "favicon.ico",
+  "favicon.png",
+  "files",
+  "groups",
+  "health_check",
+  "help",
+  "import",
+  "jwt",
+  "login",
+  "oauth",
+  "profile",
+  "projects",
+  "public",
+  "robots.txt",
+  "s",
+  "search",
+  "sitemap",
+  "sitemap.xml",
+  "sitemap.xml.gz",
+  "slash-command-logo.png",
+  "snippets",
+  "unsubscribes",
+  "uploads",
+  "users",
+  "v2",
+}
```

Comment:
I'm tempted to say that we shouldn't bother with these lists -- they're not going to be stable and they're almost certainly already incomplete/will go out of sync with GitLab over time.